sql: crdb_internal.reset_multi_region_zone_configs_for_database causes concurrent txn use detected

roachtest.sqlsmith/setup=rand-tables/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4907623&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4907623&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations) on release-21.2 @ [0f0029653c25772d09adae5be308ce0c45f84f0a](https://github.com/cockroachdb/cockroach/commits/0f0029653c25772d09adae5be308ce0c45f84f0a):


```
The test failed on branch=release-21.2, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:223,sqlsmith.go:256,test_runner.go:777: error: pq: internal error: crdb_internal.reset_multi_region_zone_configs_for_database(): concurrent txn use detected. ba: Get [/Table/3/1/2107461607/2/1,/Min), [txn: 34c72339], [can-forward-ts]
		stmt:
		WITH
			with_34293 (col_190415)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(NULL),
								('00:16:32.649786+04:54:00':::TIMETZ),
								(('10:19:57.591189-07:27:00':::TIMETZ::TIMETZ - '290 years':::INTERVAL::INTERVAL)::TIMETZ)
						)
							AS tab_87357 (col_190415)
					UNION
						SELECT
							*
						FROM
							(VALUES ('06:21:22.028838+00:04:00':::TIMETZ), (NULL), ('21:03:02.373314+12:30:00':::TIMETZ))
								AS tab_87358 (col_190416)
				),
			with_34294 (col_190417) AS (SELECT * FROM (VALUES (NULL)) AS tab_87359 (col_190417))
		SELECT
			tab_87360.col3_4 AS col_190418,
			tab_87360.col3_2 AS col_190419,
			NULL AS col_190420,
			tab_87360.col3_17 AS col_190421,
			tab_87360.col3_11 AS col_190422
		FROM
			defaultdb.public.table3@table3_col3_8_key AS tab_87360, with_34293 AS cte_ref_9835
		WHERE
			(NOT crdb_internal.reset_multi_region_zone_configs_for_database(4010021785958174695:::INT8::INT8)::BOOL)
		LIMIT
			93:::INT8;
```
<details><summary>Reproduce</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-15677

--- a/pkg/sql/sem/builtins/builtins.go
+++ b/pkg/sql/sem/builtins/builtins.go
@@ -6358,7 +6358,10 @@ the locality flag on node startup. Returns an error if no region is set.`,
 		},
 	),
 	builtinconstants.DefaultToDatabasePrimaryRegionBuiltinName: makeBuiltin(
-		tree.FunctionProperties{Category: builtinconstants.CategoryMultiRegion},
+		tree.FunctionProperties{
+			Category:         builtinconstants.CategoryMultiRegion,
+			DistsqlBlocklist: true, // applicable only on the gateway
+		},
 		stringOverload1(
 			func(evalCtx *eval.Context, s string) (tree.Datum, error) {
 				regionConfig, err := evalCtx.Regions.CurrentDatabaseRegionConfig(evalCtx.Context)
@@ -6386,7 +6389,10 @@ the locality flag on node startup. Returns an error if no region is set.`,
 		),
 	),
 	builtinconstants.RehomeRowBuiltinName: makeBuiltin(
-		tree.FunctionProperties{Category: builtinconstants.CategoryMultiRegion},
+		tree.FunctionProperties{
+			Category:         builtinconstants.CategoryMultiRegion,
+			DistsqlBlocklist: true, // applicable only on the gateway
+		},
 		tree.Overload{
 			Types:      tree.ArgTypes{},
 			ReturnType: tree.FixedReturnType(types.String),
@@ -6422,7 +6428,10 @@ the locality flag on node startup. Returns an error if no region is set.`,
 		},
 	),
 	"crdb_internal.validate_multi_region_zone_configs": makeBuiltin(
-		tree.FunctionProperties{Category: builtinconstants.CategoryMultiRegion},
+		tree.FunctionProperties{
+			Category:         builtinconstants.CategoryMultiRegion,
+			DistsqlBlocklist: true, // applicable only on the gateway
+		},
 		tree.Overload{
 			Types:      tree.ArgTypes{},
 			ReturnType: tree.FixedReturnType(types.Bool),
@@ -6442,7 +6451,10 @@ the locality flag on node startup. Returns an error if no region is set.`,
 		},
 	),
 	"crdb_internal.reset_multi_region_zone_configs_for_table": makeBuiltin(
-		tree.FunctionProperties{Category: builtinconstants.CategoryMultiRegion},
+		tree.FunctionProperties{
+			Category:         builtinconstants.CategoryMultiRegion,
+			DistsqlBlocklist: true, // applicable only on the gateway
+		},
 		tree.Overload{
 			Types:      tree.ArgTypes{{"id", types.Int}},
 			ReturnType: tree.FixedReturnType(types.Bool),
@@ -6464,7 +6476,10 @@ table.`,
 		},
 	),
 	"crdb_internal.reset_multi_region_zone_configs_for_database": makeBuiltin(
-		tree.FunctionProperties{Category: builtinconstants.CategoryMultiRegion},
+		tree.FunctionProperties{
+			Category:         builtinconstants.CategoryMultiRegion,
+			DistsqlBlocklist: true, // applicable only on the gateway
+		},
 		tree.Overload{
 			Types:      tree.ArgTypes{{"id", types.Int}},
 			ReturnType: tree.FixedReturnType(types.Bool),
@@ -6486,7 +6501,10 @@ enabled.`,
 		},
 	),
 	"crdb_internal.filter_multiregion_fields_from_zone_config_sql": makeBuiltin(
-		tree.FunctionProperties{Category: builtinconstants.CategoryMultiRegion},
+		tree.FunctionProperties{
+			Category:         builtinconstants.CategoryMultiRegion,
+			DistsqlBlocklist: true, // applicable only on the gateway
+		},
 		stringOverload1(
 			func(evalCtx *eval.Context, s string) (tree.Datum, error) {
 				stmt, err := parser.ParseOne(s)

